### PR TITLE
Create Design Doc For Plugin Redesign

### PR DESCRIPTION
**Pull request by topocount**
test plan: N/A

### CODE REVIEW

**Review by befitsandpiper — APPROVED** [review]

**Comment by elihanover** [comment]
Future considerations might include:
1. Letting plugins read from other plugins
2. Instances implementing the plugin interface

This could allow:
1. Plugins reading from other plugins
2. Plugins reading from instances
3. Instances reading from instances

**Review by elihanover — APPROVED** [review]
Future considerations might include:
1. Letting plugins read from other plugins
2. Instances implementing the plugin interface

This could allow:
1. Plugins reading from other plugins
2. Plugins reading from instances
3. Instances reading from instances

Question:
- This seems more like a redesign of the data sources that plugins read from than the plugins themselves.  Will this become a larger design doc for the grander plugin refactor?

File: design-docs/2021/q1/plugin-redesign.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+## Overview
+
+SourceCred plugins are almost completely beholden to their datasources.
+A local cache exists, but its only real purpose is to provide a local source of data for graph creation.
+Plugin caches cannot stand by themselves for too long and are easily blown out by upstream changes.
+Additionally, caches are only built in environments local to graph creation.
+They cannot easily be shared over a network, or provide any real source of truth outside of the API.
+
+@hammadj has designed an new [plugin model] that relies upon a `DataProvider` and `DataArchive` as a persistent mirror of data from a specific platform.
```
Comment (hammadj) [inline design-docs/2021/q1/plugin-redesign.md:17]: Should update the terminology here to reflect the latest naming we agreed on, `Fetcher` and `DataProvider` respectively

Comment (hammadj) [inline design-docs/2021/q1/plugin-redesign.md:17]: Should update the terminology here to reflect the latest naming we agreed on, Fetcher and DataProvider respectively

File: design-docs/2021/q1/plugin-redesign.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+This is a significant improvement over the existing model described above because it allows for the persistence of cred data independent of a platform ability (or willingness) to provide the raw information needed to build cred data over time.
+![snapshot of plugin model]
+
+[plugin model]: https://lucid.app/lucidchart/591bf744-fdbe-4aa2-a0c5-4be90f436a71/view?page=0_0#
+[snapshot of plugin model]: https://cdn.discordapp.com/attachments/714955718788382750/806248646625984542/Blank_diagram_-_Page_1.png
+
+## Informal Specification
```
Comment (hammadj) [inline design-docs/2021/q1/plugin-redesign.md:24]: Should add a link to the HackMD as well: https://hackmd.io/@nnpZD6M7TlicIVX_NaLGlA/Sk_Ye8oxO/edit

**Review by hammadj — COMMENTED** [review]
Looks good, just need to update the naming to match what we decided in the call on Friday

**Comment by topocount** [comment]
> Question:
>
> This seems more like a redesign of the data sources that plugins read from than the plugins themselves.  Will this become a larger design doc for the grander plugin refactor?

Thanks for these thoughts @elihanover!

First, to directly address your question: The logic model here is currently internal to plugins. By pulling this out, we are able to more readily implement cool stuff like perspectives on data and plugins that consume the same data for different purposes. To me, this _is_ the plugin refactor, since this condenses a plugin's responsiblities down from "(1) fetch data, then (2) organize data, then (3) store data, then (4) build graph from data" to "(1) build graph from data".

I think this enables some cool composability with raw api data, which is much more immediately important as far as reproducability of cred data goes.

Composability of plugins and instances is not something I've considered up to
this point, but it's an interesting idea, for sure.

**Pull request merged**